sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -639,6 +640,23 @@ object HiveTypeCoercion {
   }
 
   /**
+   * Turns Add/Subtract of DateType/TimestampType and CalendarIntervalType to TimeAdd/TimeSub
+   */
+  object DateTimeOperations extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+      // Skip nodes who's children have not been resolved yet.
+      case e if !e.childrenResolved => e
+
+      case Add(l @ CalendarIntervalType(), r) if r.dataType != CalendarIntervalType =>
+        Cast(TimeAdd(Cast(r, TimestampType), l), r.dataType)
```

Comment:
my comment disappeared since this part changed, so posting again.

we should avoid implicitly casting any type -- just do it for string/timestamp/date.